### PR TITLE
Infer class properties from methods and not just constructors in .js files

### PR DESCRIPTION
**Pull request by mhegazy**
Properties inferred from assignments in methods get `| undefined` in their type.

### CODE REVIEW

File: src/compiler/checker.ts [hunk 1]
```diff
@@ -3472,25 +3472,41 @@ namespace ts {
             return undefined;
         }
 
-        // Return the inferred type for a variable, parameter, or property declaration
-        function getTypeForJSSpecialPropertyDeclaration(declaration: Declaration): Type {
-            const expression = declaration.kind === SyntaxKind.BinaryExpression ? <BinaryExpression>declaration :
-                declaration.kind === SyntaxKind.PropertyAccessExpression ? <BinaryExpression>getAncestor(declaration, SyntaxKind.BinaryExpression) :
-                undefined;
+        function getWidendedTypeFromJSSpecialPropetyDeclarations(symbol: Symbol) {
```
Comment (yuit) [inline src/compiler/checker.ts]: Question: this function is not just for JS right ?( the name got me confused a bit)
Nit: would be great to see JSDoc comment...it is not clear from function name what it try to do

Comment (mhegazy) [inline src/compiler/checker.ts]: the binder only creates these declarations in a .js file.

File: tests/cases/conformance/salsa/inferringClassMembersFromAssignments.ts [hunk 2]
```diff
@@ -0,0 +1,79 @@
+// @out: output.js
+// @allowJs: true
+// @strictNullChecks: true
+
+// @filename: a.js
+class C {
+    constructor() {
+        if (Math.random()) {
+            this.inConstructor = 0;
```
Comment (sandersn) [inline tests/cases/conformance/salsa/inferringClassMembersFromAssignments.ts]: can you add another property `this.inBoth` that gets `this.inBoth = 0` in the constructor and `this.inBoth = 'string'` in `method()`. The resulting type should be `number | string` I think.

File: src/compiler/binder.ts [hunk 3]
```diff
@@ -2238,23 +2238,31 @@ namespace ts {
...
+                case SyntaxKind.Constructor:
+                case SyntaxKind.MethodDeclaration:
+                case SyntaxKind.GetAccessor:
+                case SyntaxKind.SetAccessor:
+                    // this.foo assignment in a JavaScript class
+                    // Bind this property to the containing class
+                    const containingClass = container.parent;
+                    const symbol = hasModifier(container, ModifierFlags.Static)
+                        ? declareSymbol(containingClass.symbol.exports, containingClass.symbol, node, SymbolFlags.Property, SymbolFlags.None)
+                        : declareSymbol(containingClass.symbol.members, containingClass.symbol, node, SymbolFlags.Property, SymbolFlags.None);
```
Comment (mhegazy) [inline src/compiler/binder.ts]: done.

File: tests/cases/conformance/salsa/inferringClassMembersFromAssignments.ts (hunk 2 above)
Comment (mhegazy) [inline tests/cases/conformance/salsa/inferringClassMembersFromAssignments.ts]: done.

File: src/compiler/checker.ts [hunk 4]
```diff
@@ -3472,7 +3472,7 @@ namespace ts {
             return undefined;
         }
 
-        function getWidendedTypeFromJSSpecialPropetyDeclarations(symbol: Symbol) {
+        function getWidenedTypeFromJSSpecialPropetyDeclarations(symbol: Symbol) {
```
Comment (sandersn) [inline src/compiler/checker.ts]: typo:Propety &rarr; Property

Comment (mhegazy) [inline src/compiler/checker.ts]: man.. i need a spell checker :D

File: src/compiler/binder.ts [hunk 5]
```diff
@@ -2238,23 +2238,28 @@ namespace ts {
...
+                    break;
+
+                case SyntaxKind.Constructor:
+                case SyntaxKind.MethodDeclaration:
+                case SyntaxKind.GetAccessor:
+                case SyntaxKind.SetAccessor:
+                    // this.foo assignment in a JavaScript class
+                    // Bind this property to the containing class
+                    const containingClass = container.parent;
+                    const symbol = declareSymbol(hasModifier(container, ModifierFlags.Static) ? containingClass.symbol.exports : containingClass.symbol.members, containingClass.symbol, node, SymbolFlags.Property, SymbolFlags.None);
```
Comment (rbuckton) [inline src/compiler/binder.ts]: Do we care if the member has a dynamic name or preserving control flow? That's part of what is lost by moving away from calling `bindPropertyOrMethodOrAccessor`.

File: src/compiler/checker.ts (hunk 4 above)
Comment (rbuckton) [inline src/compiler/checker.ts]: At least you fixed `Widended` -> `Widened` :)

Comment (mhegazy) [inline src/compiler/checker.ts]: yup.. baby steps

File: src/compiler/binder.ts (hunk 5 above)
Comment (mhegazy) [inline src/compiler/binder.ts]: We only call this on property expressions, so no reason to get dynamic names here.

**Pull request merged**